### ISSUE TITLE
Import runfile() produces errors when using a virtualenv

### ISSUE THREAD

**Issue by ewjoachim** [spyder-ide/spyder#2846]
Hi,

So I guess I've traced this error from symptom to potential solution :

### Steps to see the problem :
1. Install PyQt4 without pip (for example, with Ubuntu ``sudo apt-get install python-qt4``)
(note : qt is hellish to install properly via pip)
2. ``pip install spyder`` (note : I need the latest version that solves a blocking bug, so apt-get won't do)(version 2.3.8 at the time of writing)
3. Open Spyder and configure it to use a different Python interpreter, for example a virtualenv's python without --use-global-site because this may introduce a **lot** of bugs.
4. Open file, try to run code via ``F5``

### What I expect :
Code runs.

### What actually happens :
Fails due to ``ImportError`` on ``runfile``

### Reasons why this happens :
In sitecustomize, among other things, 3 things happen :

#### Python path
Thanks to the PYTHONPATH env variable set by [``pythonshell.py``](https://github.com/spyder-ide/spyder/blob/master/spyderlib/widgets/externalshell/pythonshell.py#L501), we have the ``/spyderlib/widgets/externalshell/`` in the pythonpath. However, we need to have the whole ``spyderlib`` in order to load the monitor. So we [add the folder](https://github.com/spyder-ide/spyder/blob/master/spyderlib/widgets/externalshell/sitecustomize.py#L132) containing ``spyderlib`` to the python path. Under Ubuntu 14.4 with latest pip, and python2.7 this is ``/usr/local/lib/python2.7/dist-packages/``.

#### Monitor
In the part that [loads the monitor](https://github.com/spyder-ide/spyder/blob/master/spyderlib/widgets/externalshell/sitecustomize.py#L229), there's an ImportError that is silenced because that's how [Python works](https://docs.python.org/2/library/site.html) ("If this import fails with an ImportError exception, it is silently ignored")

#### Cleaning
(we [remove](https://github.com/spyder-ide/spyder/blob/master/spyderlib/widgets/externalshell/sitecustomize.py#L606) ``/usr/local/lib/python2.7/dist-packages/`` from ``sys.path`` of course, otherwise, that would completely break the user's virtualenv)

When explored, the import error of the Monitor part above traces to [the import of ``sip``](https://github.com/spyder-ide/spyder/blob/master/spyderlib/qt/__init__.py#L23) and it fails because SIP and PyQt4 were installed using apt-get, and they were placed into ``/usr/lib/python2.7/dist-packages/`` (note : ``usr/lib`` and **not** ``/usr/local/lib``) which is not present in our current ``sys.path``

### Workaround

Once I had understood all of this, I got it to work simply by creating a symlink of PyQt4 and sip in the right folder :
```bash
cd /usr/local/lib/python2.7/dist_package
ln -s  /usr/lib/python2.7/dist_package/sip.so sip.so
ln -s  /usr/lib/python2.7/dist_package/PyQt4 PyQt4
```

### Leads for a real fix

I think what's really wrong here is either:

1. that we need the whole ``spyderlib`` to be loaded in the external python
2. or that we assume that ``spyderlib`` and ``PyQt4`` are installed in the same physical directory (all we know is that they're both in a directory present in the internal python's sys.path)

To fix 1. requires a knowledge of spyder that I don't have.
Fixing 2 can actually seem easier, but there might be tricky situations (because until we load it, we don't know in advance if we're going to use ``PyQt4`` or ``PySide`` for example). Maybe we need to store the location of ``sip`` / ``PyQt4`` / ``PySide`` from the internal python and give it to the external python for it to use **only** during launch time.

**Comment by ewjoachim** [comment]
Side note : I'd be not surprised that this would be the cause for #2844 too.

**Comment by ewjoachim** [comment]
@ccordoba12 : Just to get a better understanding : where can I find the rationale behind choosing whether an issue is a bug or an enhacement ?

**Comment by ewjoachim** [comment]
(And if I were to write a PR on that, what would be the best way according to you ?)

**Comment by goanpeca** [comment]
Hi @ewjoachim, we do not have any documentation regarding that (Bugs, enhancement), but I guess this qualifies as a bug... and needs to be addressed.

Now, **Spyder** is meant to be used for scientific applications involving python, and pip has not been very good handling the scientific stack (numpy, scipy, matplotlib....). You also mention you are using **virtualenv**, which again does not work very well when you want to isolate other libraries (like pyqt in this case). Having said that I would recommend that if you plan to use Spyder you use **[conda](http://conda.pydata.org/docs/)** (and possibly conda environments).

**Comment by ewjoachim** [comment]
Ah, it's probably a question of different approaches... We're trying to have a compatible stack between IT and Research (both using Python), and having everyone using pip / virtualenv greatly helps us when it comes to integrating Research's work and putting it into production. Now if you have good descriptions of workflow involving half of the team using Pip/virtualenv, half using Conda, and all the code playing nicely together when put in the same context, I'd be happy to learn a thing or two :)

Up until now, I feel like I've seen little documentation on the subject and we've had problems in the past, so that's why we're asking everyone to use pip / virtualenv. And it feels like the problem it involves are usually easier to solves than the ones we'de get with a completely different stack from research/development to production... But again, I'm taking any advice !

**Comment by goanpeca** [comment]
Not sure what IT implies, but if the idea was to have both teams on the same stack and you are using things that depend on non python stuff I (personal opinion...) would advice you to have them both using Conda, I have never looked back since I did the change and I am doing both development (IT I guess you can call it) and research (though limited these days) work.

>Now if you have good descriptions of workflow involving half of the team using 
Pip+virtualenv does not completely solve the hard dependency problems with non python dependencies (often the case in research), and in that case I would choose the tool that gives more coverage. And if there is not a conda package available, then you can always install pip packages inside a conda environment

Pip/virtualenv, half using Conda, and all the code playing nicely together when put in the same context, I'd be happy to learn a thing or two :)

This wont work cause **conda** ann **virtualenv** do not work well together, when using **conda**, **virtualenv** is no longer needed, cause **conda environments** are way superior to **virtualenv environments**. So what I am suggesting is using conda/conda-environments/pip.  

You can create an environment with conda doing

```bash
conda create -n envname python=2.7 numpy=1.9
source activate envname
pip install packagename
```

**Comment by ccordoba12** [comment]
This will be fixed by PR #2878 for the IPython console (if you have `ipykernel` installed in your virtualenv).

We can't do the same for the Python console, but at least I'd say this is progress :-)
